### ISSUE TITLE
[RFC] TPA Settings editor flow integration

### ISSUE THREAD

**Issue by yavorsky** [wix/yoshi#2725]
# TPA Settings integration

Here I'll try to explain related to tpa-settings package integraiton to our flow.

Editor Flow should provide best practices for pre-configured features under the hood and templates we are generating.

A lot of different companies using the different technologies to achieve the same goal. We want to change this. By maintaining single library we can bring much more values.

[tpa-settings](https://github.com/wix-private/tpa-settings) is a library created by @tuchk4 to improve development velocity during working on Out of iFrame apps, especially on Settings panel part.

In general, it's helping working with 2 kinds of data:
- Settings
- Styles

#### Settings
Some widget configuration data available mostly in application code and not related directly to css values.
*For example, some placholder text:*
<img width="290" alt="Screenshot 2020-07-16 at 12 10 25" src="https://user-images.githubusercontent.com/1521229/87652847-65154480-c75d-11ea-8188-fdd5f087341c.png">

*Or whether or not showing some extra elements:*
<img width="278" alt="Screenshot 2020-07-16 at 12 08 15" src="https://user-images.githubusercontent.com/1521229/87652908-79594180-c75d-11ea-88f8-656753f88f81.png">

#### Styles
Something related to css values (bg color, font, text color, etc)
*Font example:*
<img width="540" alt="Screenshot 2020-07-16 at 12 12 25" src="https://user-images.githubusercontent.com/1521229/87653043-a73e8600-c75d-11ea-808f-d3b3213c900f.png">

The basic idea of having specific library for it - it's not easy to handle it. It's a lot of pitfalls and complexities you can face during tpa settings development. Moreover, there are some good practices for all settings panels, so we can reduce boilerplate on the user's side and provide some ready-to-use solutions like blocks.

We want also to have color and font pickers to be preconfigured and looks like simple component.
Moreover, since editor flow supports only TS, we want this lib also be fully-typed.

That's why tpa-settings matching almost all of our needs.

## Extending editor flow file-structure convention

Currently, we've added tpa-settings to the editor flow template and it works cool. The next step is to integrate it more with editor flow on 'magic' level.

The proposed solution could contains of few parts and integration should be gradual.

The next goal after adding it to the editor flow as a raw implementation will be reducing boilerplate user have to maintain.

Currently, we have a convention for `Widget` (module that exports React Component and being rendered in Editor and Viewer parts), `Settings` - module with react component with settings panel and `controller` - a function that being collected into a `viewerScript`.

By extending a file-structure API on early stages, we will have less projects using this boilerplate and migration path will be much easier. On the other hand, we should implement best approach from the start and prevent cases when we are going to rename files, support 2 kinds of the same thing, etc.

### Pass values to controller:
Instead of importing and using some boilerplate functions, we can do it under the hood:

*controller*
```ts
const createController: CreateControllerFn = async ({
  flowAPI,
}) => {
  const { setProps } = flowAPI.controllerConfig;

  const publicData = flowAPI.controllerConfig.config.publicData.COMPONENT || {};
  const settings = getSettingsValues(publicData, componentSettings);

  return {
    async pageReady() {
      setProps({
        greetingsText: settings.greetingsText,
      });
    },
    updateConfig($w, config) {
      const updatedPublicData = config.publicData.COMPONENT || {};
      const updatedSettings = getSettingsValues(
        updatedPublicData,
        componentSettings,
      );

      setProps({
        greetingsText: updatedSettings.greetingsText,
      });
    },
  };
};
```

Let's analyse the controller structure above:
- `config.publicData.COMPONENT`, `getSettingsValues` - boilerplate. We can do it under the scenes and pass `settings` object. `componentSettings` will be taken by the convention.
- In `updateConfig` we just taking updated values and passing it via setProps. By having a wrapper for controller, we can easily do the same under the hood. We can verify which fields were changed and set it to prevent extra `setProps` calls.
- Since we are already passing translations / experiments under the scenes, we can also do the same for settings. (If we'll see some performance issues with this approach, we can talk to platform guys or provide some declarative way to specify which settings you want to pass).

After these changes controller should look like almost like:
*controller*
```ts
const createController: CreateControllerFn = async ({
  flowAPI,
}) => {
 const { componentSettings } = flowAPI;

  // Do something with settings in controller if needed:
  console.log(componentSettings.greetingsText)

  return {
    async pageReady() {
      setProps({
         // No need to pass anything. `setProps` can be avoided
      });
    },
};
```

In `Widget` part `SettingsConsumer` will be available.
*Widget.ts*
```tsx
<SettingsConsumer>
        {(settings) => (
           <span>{settings.get(componentSettings.greetingsText)}</span>
        )}
</SettingsConsumer>
```
To prevent duplication for `componentSettings` Consumer, we can merge settings and settingsValues we will use for getter into a single Consumer..

### Automatically using `componentSettings.ts` and `componentStyles.ts`
(we can find a better naming for these files)
In practice, these files contain values we are supporting.
Check these examples: [componentSettings](https://github.com/wix-private/tpa-settings/blob/master/packages/tpa-settings/examples/appSettings.ts) and [appStyles](https://github.com/wix-private/tpa-settings/blob/master/packages/tpa-settings/examples/appStyles.ts).
Basically it's a values configuration. Here we can set default value, key and type of the style.

So in these files we have enough info to:
- Consume in controller settings we have configured
- Be able to pass these values via provider/consumer instead of asking user to directly import it.
- Take some advantage of code generation and allow users to do `styles.get('textColor')` instead of `styles.get(componentStyles.textColor)`.
- Provide featues like Mobile settings and multilingual under the hood.

The current approach is to take manually values and use it:
*Settings.ts*
```ts
import { componentStyles } from '../../componentStyles'; // taking a value

<StylesConsumer>
   {(styles) => (
      <OpacityColorPicker
          title="Background Color"
          dataHook="design-tab-color-picker"
          value={styles.get(componentStyles.backgroundColor)}
          onChange={(value) => {
            styles.set(componentStyles.backgroundColor, value);
          }}
        />
   )}
```

If we'll take componentStyles on the flow level, we can get rid of some things here. Just passing it via context and generating types for should work for us:

```ts
import { componentStyles } from '../../componentStyles'; // taking a value

<ComponentStylesConsumer>
     {(componentStyles) => (
        <OpacityColorPicker
            title="Background Color"
            dataHook="design-tab-color-picker"
            valueType={componentStyles.backgroundColor} // We'll take the styles from the context and use the value from here.
         />
      )}
</ComponentStylesConsumer>
```
In this case we will won't control `onChange` field, value field won't be a getter, but tpa-settings value type. It will allow us to provide some validation, but on the same time prevent use getters and setters each time.

Maybe in this case I could miss something, but would be good to know if there are any pitfalls with this approach.

In general it's pretty raw, by it would be cool to continue discussion here and extend the proposed approach with fresh ideas.

**Comment by tuchk4** [comment]
### Code generation for settings/styles types

The current way to use `OpacityColorPicker` (or any other component) with the styles from the `tpa-settings`: 

```js
<StylesConsumer>
     {(styles) => (
        <OpacityColorPicker
            title="Background Color"
            dataHook="design-tab-color-picker"
            value={componentStyles.get(styles.backgroundColor)}
            onChange={value => componentStyles.set(styles.backgroundColor, value)}
         />
      )}
</StylesConsumer>
```

As @yavorsky said:
> If we'll take componentStyles on the flow level, we can get rid of some things here. Just passing it via context and generating types for should work for us:

```js
<ComponentStylesConsumer>
     {(componentStyles) => ( // Received automatically
        <OpacityColorPicker
            title="Background Color"
            dataHook="design-tab-color-picker"
            valueType={componentStyles.backgroundColor} // We'll take the styles from the context and use the value from here.
         />
      )}
</ComponentStylesConsumer>
``` 

From my point of view, I don't think this is a significant improvement. 
I would like to keep "magic-code-generation" as less as possible and try to focus on how we can push newer React version (for the viewer part) and use **hooks** instead of consumers.

Also in this way, we are losing the `go-to-defenition` feature, by cmd+clicking on `styles.backgroundColor,`


### Passing settings to the viewer

```js
const createController: CreateControllerFn = async ({
  flowAPI,
}) => {
 const { componentSettings } = flowAPI;

  // Do something with settings in controller if needed:
  console.log(componentSettings.greetingsText)

  return {
    async pageReady() {
      setProps({
         // No need to pass anything. `setProps` can be avoided
      });
    },
};
```

> No need to pass anything. `setProps` can be avoided
Do you mean, that all settings automatically passed to the viewer?

**Comment by tuchk4** [comment]
Also, note, there is a PR at `tpa-settings` that improves the settings/styles definitions:

https://github.com/wix-private/tpa-settings/pull/42

```ts
const appSettings = createSettingsParams<{
  appLayout: 'left' | 'right';
  itemsPerPage: number;
}>({
  appLayout: {
    getDefaultValue() {
      return 'left';
    },
  },
  itemsPerPage: {
    getDefaultValue() {
      return 15;
    },
  },
});
```

As a result:
- get rid of `name | key` properties
- get rid of extra interfaces


<details>
<summary>Example of how to create settings/styles now</summary>
```ts
export type IAppSettings = ISettingsParams<{
  layout: ISettingsParam<'left' | 'right'>,
  itemsPerPage: ISettingsParams<'number'>,
}>;

const appSettings: IAppSettings = {
  layout: {
    key: 'appLayout', // key to store param in settings store
    getDefaultValue() {
      return 'left';
    },
  },
  itemsPerPage: {
    key: 'itemsPerPage', // key to store param in settings store
    getDefaultValue() {
      return 15;
    },
  },
};
```
</details>

**Comment by yavorsky** [comment]
> Do you mean, that all settings automatically passed to the viewer?

yep, so we can use the same API in Widget part and don't need to have a controller layer for it (we are doing it for experiments, translations)

**Comment by bodia-uz** [comment]
> Also note, there are APP settings and COMPONENT settings. I don't know the 100% difference between them, but know that some OOI uses both of them.

`site-search` currently uses both for example:
https://github.com/wix-private/site-search/search?q=publicData.&unscoped_q=publicData.

**Comment by shahata** [comment]
> No need to pass anything. setProps can be avoided

That's great! I think in order to make generator coherent we should fetch some data using yoshi server then and set it as props

**Comment by shahata** [comment]
I'm missing some more descriptions regarding how we are going to handle mobile/breakpoints/multilingual and how we are going to handle undo/redo. Will all this be in a separate rfc?

**Comment by ronnyrin** [comment]
Didn't we say we will stop writing RFCs in github? :-(

> Also note, there are APP settings and COMPONENT settings. I don't know the 100% difference between them, but know that some OOI uses both of them.

Yes, `APP` scope is used for keys that you want to access from different widgets like in our case, the user can configure what will be written on the add to cart button and we have the button in multiple places (product page, gallery, product widget ec) so we thought it will be cool if he will configure it once and all the widgets will display the value. `COMPONENT` scope is saved per instance of the widget so if the user added two galleries to his page, he can change the text of "Filter by" for each gallery.
This worked great until we integrated with multi-lingual because in app-settings there isn't `APP` scope so we had to migrate all of our keys to use `COMPONENT` scope.

Regarding passing all the settings behind the scenes, where the merging of the free text and translation will happen? for each free that the user can change (like "add to cart"), there is always a default value that we get from the translation. You will have to merge it somehow.

Regarding removing `onChange` from the settings component, note that we send BI on each change so maybe it's not worth it as it will always require the developer to explicit write his own `onChange` .

**Comment by shahata** [comment]
> Regarding passing all the settings behind the scenes, where the merging of the free text and translation will happen? for each free that the user can change (like "add to cart"), there is always a default value that we get from the translation. You will have to merge it somehow.

Good point! I think this means that in `componentSettings.ts`, the `getDefaultValue` should get `t` as argument.

> Regarding removing onChange from the settings component, note that we send BI on each change so maybe it's not worth it as it will always require the developer to explicit write his own onChange .

Or maybe we can get the BI event id as prop and send it from tpa-settings...

**Comment by ronnyrin** [comment]
> Good point! I think this means that in componentSettings.ts, the getDefaultValue should get t as argument.

You need it both in the settings and in the widget.

**Comment by shahata** [comment]
> You need it both in the settings and in the widget.

Isn't it used in both?

**Comment by ronnyrin** [comment]
oh, yes, i think it is.

**Comment by shahata** [comment]
so it's beautiful :)

**Comment by yavorsky** [comment]
> I'm missing some more descriptions regarding how we are going to handle mobile/breakpoints/multilingual and how we are going to handle undo/redo.

Yeah, I'll do it for tpa-settings repo as a separate RFCs.

> Good point! I think this means that in componentSettings.ts, the getDefaultValue should get t as argument.

Exactly! It will be a part of multilingual effort I think.

> Regarding removing onChange from the settings component, note that we send BI on each change so maybe it's not worth it as it will always require the developer to explicit write his own onChange .

As a option, we can have `onChanged` or something like this, so users can do some additional work here.

 > Isn't componentSettings.ts used in both?

Yep. That's why I like the concept of having settings panel as a part of Component entity near Widget and controller. We can have `componentSettings.ts` reusable between all of it and preventing same logic duplication.

**Comment by yavorsky** [comment]
> Didn't we say we will stop writing RFCs in github? :-(

I'll try to find a better solution to comment stuff for the next time!

**Comment by shahata** [comment]
> Exactly! It will be a part of multilingual effort I think.

Actually supporting translations in `getDefaultValue` is much more basic feature than multilingual, the fact that we currently don't support it is a blocker to any app, even if it doesn't support multilingual. We should probably give this high priority.

> As a option, we can have `onChanged` or something like this, so users can do some additional work here.

I would prefer to also deal with BI sending ourselves, it is an abstraction that will save a lot of boilerplate

**Comment by ranyitz** [comment]
Regarding translated default values, it has also been [request by @sarunas in this thread](https://wix.slack.com/archives/C015R0X2ND9/p1594807912012900) and I agree that it's important.
